addons/zylann.hterrain/hterrain_data.gd

```diff
@@ -1426,7 +1426,72 @@ func _import_heightmap(fpath: String, min_y: int, max_y: int, big_endian: bool)
 				f.get_16()
 
 		im.unlock()
+	elif ext == "xyz":
+		_logger.debug("detected xyz file")
+		var f := File.new()
+		var err := f.open(fpath, File.READ)
+		if err != OK:
+			return false
+
+		var line = f.get_line()
+		# f.get_8() get byte to string if \r then win ending \r\n (2) else linux ending \n (1)
+		var file_len = f.get_len()
+		var file_res = Util.integer_square_root(file_len / (line.length() + 2)) # + 2 ist /r/n should be dynamic
+		var res := get_adjusted_map_size(file_res, file_res)
+
+		var width := res
+		var height := res
+		var act_min := 10000
+		var act_max := 0
+
+		_locked = true
+
+		_logger.debug(str("Resizing terrain to ", width, "x", height, "..."))
+		resize(res, false, Vector2())
+
+		var im := get_image(CHANNEL_HEIGHT)
+		assert(im != null)
+
+		var hrange := max_y - min_y
+
+		_logger.debug("Converting to internal format...")
+
+		im.lock()
 
+		var rw := Util.min_int(res, file_res)
+		var rh := Util.min_int(res, file_res)
+
+		_logger.debug("{0} {1}".format([rw, rh]))
+
+		var fline = line.split(" ")
+		var f_x := int(fline[0])
+		var f_y := int(fline[1])
+		var x := 0
+		var y := 0
+		var gs := float(fline[2])
+		# _logger.debug(
+		# 	"x: {0}, x: {1}, {2}, {3}, {4}".format([file_res - x, x, float(fline[0]), f_x, gs])
+		# )
+		im.set_pixel(file_res - x, y, Color(gs, 0, 0))
+		while !f.eof_reached():
+			fline = f.get_csv_line(" ")
+			if fline.size() < 2: # last line is empty
+				break
+			x = (float(fline[0]) - f_x)
+			y = (float(fline[1]) - f_y)
+			gs = float(fline[2])
+			act_min = min(act_min, gs)
+			act_max = max(act_max, gs)
+			# _logger.debug(
+			# 	"x: {0}, x: {1}, {2}, {3}, {4}".format([file_res - x, x, float(fline[0]), f_x, gs])
+			# )
+			im.set_pixel(file_res - x, y, Color(gs, 0, 0))
```

Comment:
You are right since there is no real documentation about that format I can't guarrantee it either. The issue here is again, that I would need to iterate the full file and read it to make a correct calculation. All the files I have, are ordered, though. Here the issue again is, that I would need to iterate the whole file, which can take a while, without being able to notify the user.

Reply:
Iterating the whole file should be fine. It's an import process which is not used in game, and is not going to be used too often. That said there will always be a need for assumptions here because XYZ points could really be anywhere. It also comes back to my big surprise that GIS software would only give you this format, because it is a very inefficient one...